### PR TITLE
[kernel] rework opaques definitions with no type

### LINKED ISSUE #168: [anomaly] coq.env.add-const Bo _ tt (opaque constant with no type)
Apparently the vernac/declare.ml API of Coq is happy with a transparent definition with no type, but complains if it is opaque (it should infer it) see
https://github.com/coq/coq/blob/e0b8b4684eaf76f897ac708ffddbb8e4977ac754/vernac/declare.ml#L274

This limitation should disappear if `Theorem foo := body` becomes valid syntax as per https://github.com/coq/ceps/pull/42 (cc @Zimmi48)

### PR DESCRIPTION
**Pull request by gares**
This PR does two things:
- [x] improve kernel API so that one can declare an opaque constant without giving its type (Fix LPCIC/coq-elpi#168)
- [ ] provide user syntax for it as per coq/ceps#42 , CC @Zimmi48 

I open this early because I need feedback on the kernel API, in particular I could not find which invariant the generalized/instantiated type variable `'opaque` of `constant_def` was imposing. Also, most of this patch is a renaming since opaque != delayed.

### CODE REVIEW

**Comment by gares** [comment]
@ppedrot I think you are the author of the refactoring that introduces the type variable `'opaque`, I may use some help. See the `Obj.magic` in the code, which is clearly to be removed.

File: kernel/term_typing.ml [hunk 1]
```diff
@@ -176,8 +176,12 @@ let infer_declaration env (dcl : constant_entry) =
            Vars.subst_univs_level_constr usubst tj.utj_val
       in
       let def = Vars.subst_univs_level_constr usubst j.uj_val in
-      let def = Def (Mod_subst.from_val def) in
-        feedback_completion_typecheck feedback_id;
+      let def =
+        if c.const_entry_opaque then
+          Obj.magic @@ OpaqueDef ()
```
Comment (ppedrot) [inline kernel/term_typing.ml:181]: Wut?

Comment (gares) [reply kernel/term_typing.ml:181]: please, read my comments ;-)

Comment (ppedrot) [reply kernel/term_typing.ml:181]: Sorry, race condition. I was reading the code while you were writing the comment :/

**Comment by ppedrot** [comment]
Why not infer the type outside of the kernel, by the way? You're forcing the body of the constant anyways when the type is absent.

**Comment by gares** [comment]
(well, or 3 times...)

**Comment by gares** [comment]
Anyway, can you explain me why you made `OpaqueDef` take an argument? (and hence generate a monomorphic value?)

**Comment by ppedrot** [comment]
We use the parametrization to ensure that opaque proofs are correctly set into the environment for side-effects. It's probably a bit annoying to have two different types for this but you might give it a try.

**Comment by gares** [comment]
You mean this?
```diff
--- a/kernel/declarations.ml
+++ b/kernel/declarations.ml
@@ -51,10 +51,11 @@ type inline = int option
     transparent body, or an opaque one *)
 
 (* Global declarations (i.e. constants) can be either: *)
-type ('a, 'opaque) constant_def =
-  | Undef of inline                       (** a global assumption *)
-  | Def of 'a                             (** or a transparent global definition *)
-  | OpaqueDef of 'opaque                  (** or an opaque global definition *)
+type ('a, 'delayed) constant_def =
+  | Undef of inline            (** a global assumption *)
+  | Def of 'a                  (** or a transparent global definition *)
+  | OpaqueDef                  (** or an opaque global definition *)
+  | DelayedProof of 'delayed   (** or an opaque global definition *)
   | Primitive of CPrimitives.t (** or a primitive operation *)
 
 type universes =
```

**Comment by gares** [comment]
@ppedrot this should be better

**Comment by gares** [comment]
nah, still broken...

File: kernel/safe_typing.ml [hunk 2]
```diff
@@ -883,30 +885,39 @@ let export_private_constants eff senv =
   let senv = List.fold_left add_constant_aux senv bodies in
   exported, senv
 
+  let opacify_body senv cb =
+    let c = match cb.const_body with
+      | Def x -> x
+      | _ -> assert false in
+    let fc = Future.from_val (Mod_subst.force_constr c, Opaqueproof.PrivateMonomorphic Univ.ContextSet.empty) in (* TODO *)
```
Comment (gares) [inline kernel/safe_typing.ml:892]: I kind of managed to make it type, but I don't know what I should do here w.r.t. the universe part.

**Comment by gares** [comment]
> Why not infer the type outside of the kernel, by the way? You're forcing the body of the constant anyways when the type is absent.

I thought about that, but your remark also applies to regular definition (which do have an optional type, as in the syntax).

**Comment by ppedrot** [comment]
The goal of some of the previous factorization work was to bring closer the code path for axioms and Qed-terminated proofs, to ultimately move the Future part out of the kernel. I think that pushing the type inference of opaque proofs inside the kernel is a regression from that point of view, but it's probably not critical.

**Comment by gares** [comment]
Wrt this plan, I did try to have another constructor for opaque, not delayed, proofs. So we would have `Delayed | OpaqueDef`. If we go that way, then I think we get closer to moving opaque stuff out of the kernel, but this is a bit more invasive since higher layers look for the constant body (eg reduction code/tactics).

Wrt the regression, IMO the alternative is to make the type mandatory for transparent constants as well (in the entry data type) and use the one synthesized by pretyping if omitted in the concrete syntax. This would make the code more regular, IMO.

File: kernel/entries.ml [hunk 3]
```diff
@@ -78,14 +80,14 @@ type section_def_entry = {
...
   (* List of section variables *)
-  opaque_entry_secctx : Id.Set.t;
+  delayed_entry_secctx : Id.Set.t;
   (* State id on which the completion of type checking is reported *)
-  opaque_entry_feedback : Stateid.t option;
-  opaque_entry_type        : types;
-  opaque_entry_universes   : universes_entry;
+  delayed_entry_feedback : Stateid.t option;
+  delayed_entry_type        : types;
+  delayed_entry_universes   : universes_entry;
```
Comment (ejgallego) [inline kernel/entries.ml:90]: nit: Maybe and opportunity to make this `Entries.Delayed.t` ?

**Comment by Zimmi48** [comment]
> Wrt the regression, IMO the alternative is to make the type mandatory for transparent constants as well (in the entry data type) and use the one synthesized by pretyping if omitted in the concrete syntax. This would make the code more regular, IMO.

:+1:

**Comment by gares** [comment]
I lost interest in this PR, I won't remove the branch in case someone wants to revive it

**Pull request closed without merging**